Common/Securities/Future/FuturesExpiryFunctions.cs

```diff
@@ -954,6 +1220,147 @@ public class FuturesExpiryFunctions
+                        .GetEntry("usa", Futures.Energies.SingaporeMogas92UnleadedPlattsBrentCrackSpread, SecurityType.Future)
+                        .ExchangeHours
+                        .Holidays;
+
+                    while (holidays.Contains(lastBusinessDay) || !lastBusinessDay.IsCommonBusinessDay())
+                    {
+                        lastBusinessDay = FuturesExpiryUtilityFunctions.AddBusinessDays(lastBusinessDay, -1);
+                    }
+
+                    return lastBusinessDay;
+                })
+            },
+            // Dubai Crude Oil (Platts) Financial (DCB): https://www.cmegroup.com/trading/energy/crude-oil/dubai-crude-oil-calendar-swap-futures_contract_specifications.html
+            {Futures.Energies.DubaiCrudeOilPlattsFinancial, (time =>
+                {
+                    // Trading shall cease on the last London and Singapore business day of the contract month
+                    var lastBusinessDay = FuturesExpiryUtilityFunctions.NthLastBusinessDay(time, 1);
+                    var holidays = MarketHoursDatabase.FromDataFolder()
+                        .GetEntry("usa", Futures.Energies.DubaiCrudeOilPlattsFinancial, SecurityType.Future)
+                        .ExchangeHours
+                        .Holidays;
+
+                    while (holidays.Contains(lastBusinessDay) || !lastBusinessDay.IsCommonBusinessDay())
+                    {
+                        lastBusinessDay = FuturesExpiryUtilityFunctions.AddBusinessDays(lastBusinessDay, -1);
+                    }
+
+                    return lastBusinessDay;
+                })
+            },
+            // JJapan C&F Naphtha (Platts) BALMO (E6): https://www.cmegroup.com/trading/energy/refined-products/japan-naphtha-balmo-swap-futures_contract_specifications.html
+            {Futures.Energies.JapanCnFNaphthaPlattsBALMO, (time =>
+                {
+                    // Trading shall cease on the last business day of the contract month. 
+                    return FuturesExpiryUtilityFunctions.NthLastBusinessDay(time, 1);
+                })
+            },
+            // Ethanol (EH): https://www.cmegroup.com/trading/energy/ethanol/cbot-ethanol_contract_specifications.html
+            {Futures.Energies.Ethanol, (time =>
+                {
+                    // Trading terminates on 3rd business day of the contract month in "ctm"
+                    var currentDay = time;
+                    var daysCounted = time.IsCommonBusinessDay() ? 1 : 0;
+                    var i = 0;
+                    var holidays = MarketHoursDatabase.FromDataFolder()
+                        .GetEntry("usa", Futures.Energies.Ethanol, SecurityType.Future)
+                        .ExchangeHours
+                        .Holidays;
+                    
+                    while (daysCounted < 3)
+                    {
+                        if (holidays.Contains(currentDay) || USHoliday.Dates.Contains(currentDay))
```

Comment:
Why holidays in MHDB doesn't contain the holidays in `USHoliday.Dates`?

Reply:
The asset continues trading on some of the dates in the USHoliday.Dates list, but the behavior of the last trade date depends on some of the holidays (2021 New Years day) not listed in MHDB